drivers/hmis/lib/tasks/one_time_migration_20240303.rake

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+desc 'One time data migration to transform link_ids into valid identifiers'
+# rails driver:hmis:migrate_hmis_link_ids
+task migrate_hmis_link_ids: [:environment] do
+  Hmis::Form::Definition.transaction do
+    OneTimeMigration20230303.new.perform
+    # raise ActiveRecord::Rollback
+  end
+end
+
+class OneTimeMigration20230303
+  def perform
+    update_form_processors
+    update_form_definitions
+  end
+
+  # Migrate LinkIDs for all files in drivers/hmis/lib/form_data. Write changes to files.
+  # Doesn't catch everything, a few things need to be manually fixed (client-id and veteran-status)
+  def update_form_data_files
+    Dir.glob("#{HmisUtil::JsonForms::DATA_DIR}/**/*.json").each do |file_path|
+      puts "Transforming #{file_path}..."
+      file = File.read(file_path)
+      parsed = JSON.parse(file)
+
+      if parsed.is_a?(Array) # patch files are arrays
+        parsed.each { |item| transform_definition(item) }
+      else
+        transform_definition(parsed)
+      end
+
+      File.open(file_path, 'w') do |f|
+        f.write(JSON.pretty_generate(parsed))
+      end
+    end
+  end
+
+  def update_form_processors
+    Hmis::Form::FormProcessor.where.not(values: nil).find_each do |processor|
+      processor.values = processor.values.transform_keys do |key|
+        transform_identifier(key)
+      end
+      processor.save!
+    end
+  end
+
+  def update_form_definitions
+    Hmis::Form::Definition.find_each do |definition|
+      next unless definition.valid?
```

Comment:
Maybe we should log if we encounter these, or check beforehand if any of the definitions are invalid and fix them before running this?

Reply:
Good idea, yeah I added a more robust validation check. I had this mostly for my local dev env that has some garbage form definitions